modules/platforms/cpp/thin-client/src/impl/transactions/transactions_impl.h

```diff
@@ -0,0 +1,264 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef _IGNITE_IMPL_THIN_TRANSACTIONS_IMPL
+#define _IGNITE_IMPL_THIN_TRANSACTIONS_IMPL
+
+#include "impl/data_router.h"
+#include <ignite/common/fixed_size_array.h>
+#include "ignite/thin/transactions/transaction_consts.h"
+
+namespace ignite
+{
+    namespace impl
+    {
+        namespace thin
+        {
+            namespace transactions
+            {
+                /* Forward declaration. */
+                class TransactionsImpl;
+
+                /* Forward declaration. */
+                class TransactionImpl;
+
+                typedef ignite::common::concurrent::SharedPointer<TransactionsImpl> SP_TransactionsImpl;
+                typedef ignite::common::concurrent::SharedPointer<TransactionImpl> SP_TransactionImpl;
+
+                /**
+                 * Ignite transactions implementation.
+                 *
+                 * This is an entry point for Thin C++ Ignite transactions.
+                 */
+                class TransactionImpl
+                {
+                public:
+                    /**
+                     * Constructor.
+                     *
+                     * @param txImpl Transactions implementation.
+                     * @param txid Transaction Id.
+                     * @param concurrency Transaction concurrency.
+                     * @param isolation Transaction isolation.
+                     * @param timeout Transaction timeout.
+                     * @param size Number of entries participating in transaction (may be approximate).
+                     */
+                    TransactionImpl(
+                            TransactionsImpl& txImpl,
+                            int32_t txid,
+                            ignite::thin::transactions::TransactionConcurrency::Type concurrency,
+                            ignite::thin::transactions::TransactionIsolation::Type isolation,
+                            int64_t timeout,
+                            int32_t size) :
+                        txs(txImpl),
+                        txId(txid),
+                        concurrency(concurrency),
+                        isolation(isolation),
+                        timeout(timeout),
+                        txSize(size),
+                        closed(false)
+                    {}
+                    
+                    /**
+                     * Destructor.
+                     */
+                    ~TransactionImpl() {}
```

Comment:
This will cause memory leak if transaction object is destroyed on C++ side without calling `Close()`. Destructor should always release all resources it holds.